### PR TITLE
Md5 enhance impl and test

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/1619

Updating Hash design to fix:
multi-thread support by not using a singleton handle for the hash

### CODE REVIEW

File: sdk/core/azure-core/CMakeLists.txt [hunk 1]
```diff
@@ -47,7 +47,9 @@ set(
     ${WIN_TRANSPORT_ADAPTER_INC}
     inc/azure/core/credentials/credentials.hpp
     inc/azure/core/credentials/token_credential_options.hpp
+    inc/azure/core/cryptography/hash_base.hpp
     inc/azure/core/cryptography/hash.hpp
+    inc/azure/core/cryptography/md5.hpp
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CMakeLists.txt:52]: Add explicitly to the simplified header and test.

Comment (ahsonkhan) [reply sdk/core/azure-core/CMakeLists.txt:52]: That said, I recommend reverting this change and keeping everything in the single `hash.hpp` header file. People on the C++ team had advised **against** too granular header files (for example one per type), and were in favor of fewer, bigger header files (unless you expect that header would end up so big that compiler throughput would be measurably damaged by doing so). This is also easier for the end user and reduces a bunch of indirection and links within the generated docs.

Comment (ahsonkhan) [reply sdk/core/azure-core/CMakeLists.txt:52]: More importantly, it is difficult for me to see and understand what exactly is changing because of this file split, in this PR. It makes reviewing code and logic difficult (I end up reviewing other things like doc comments, etc.). If you feel strongly about splitting it, please do it in a separate PR.

**Review by ahsonkhan — CHANGES_REQUESTED** [review]
Please avoid the header file split, at least for now, so the change is easier to review.

**Comment by ahsonkhan** [comment]
We also need to add tests for the edge cases listed in the issue.

File: sdk/core/azure-core/src/cryptography/md5.cpp [hunk 2]
```diff
@@ -16,100 +16,66 @@
...
 
-  void Md5Hash::OnAppend(const uint8_t* data, std::size_t length)
-  {
-    _detail::Md5HashContext* md5Context = static_cast<_detail::Md5HashContext*>(m_md5Context);
+class Md5BCrypt : public Azure::Core::Cryptography::Hash {
+private:
+  std::string m_buffer;
+  BCRYPT_HASH_HANDLE m_hashHandle = nullptr;
+  std::size_t m_hashLength = 0;
+  AlgorithmProviderInstance m_algorithmProviderInstance;
```
Comment (Jinming-Hu) [inline sdk/core/azure-core/src/cryptography/md5.cpp:73]: You must NOT make it a global static variable. `BCryptOpenAlgorithmProvider` cannot be called during dll initialization.

**Review by Jinming-Hu — CHANGES_REQUESTED** [review]

File: sdk/core/azure-core/src/cryptography/md5.cpp (hunk 2 above)
Comment (Jinming-Hu) [reply sdk/core/azure-core/src/cryptography/md5.cpp:73]: And we should add a comment for that to remind future developers.

Comment (vhvb1989) [reply sdk/core/azure-core/src/cryptography/md5.cpp:73]: Is this a global static?

Isn't it just a private field from an instance?

btw, we have the dll header to create global statics just fine  @antkmsft  added that.  But, I am confused here since I don't exepect it to be static global, but one per instance

Comment (Jinming-Hu) [reply sdk/core/azure-core/src/cryptography/md5.cpp:73]: static variable defined in a function body is fine. Otherwise, it's global.

File: sdk/core/azure-core/CMakeLists.txt (hunk 1 above)
Comment (vhvb1989) [reply sdk/core/azure-core/CMakeLists.txt:52]: > keeping everything in the single hash.hpp

As we keep adding Hashing algorithms, we should be able to include all of them or just the one we want.
We will need to add SHA next and maybe others. Let's make it easy for developers and also for maintainers :)

Comment (vhvb1989) [reply sdk/core/azure-core/CMakeLists.txt:52]: > More importantly, it is difficult for me to see and understand what exactly is changing

I can show/tell the main changes and explain any question if you want.  I wouldn't really like to revert XD,  I would prefer to let it up to whatever you'd like to do :)

**Comment by vhvb1989** [comment]
> We also need to add tests for the edge cases listed in the issue.

Tests are already there for the questions.  
We throw invalid args and runtime error for the edge cases.

All I did was remove the static field from the Windows implementation and the void* from the general abstraction in favor or a unique_ptr<Hash>

File: sdk/core/azure-core/src/cryptography/md5.cpp (hunk 2 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/cryptography/md5.cpp:73]: > static variable defined in a function body is fine. Otherwise, it's global.

class A {
private:
  std::string m_someVar;
}

Are you telling me that `m_someVar` in previous example is global?

Comment (Jinming-Hu) [reply sdk/core/azure-core/src/cryptography/md5.cpp:73]: It's not a `global` variable per se. But it has the same lifecycle as global variables.

Comment (vhvb1989) [reply sdk/core/azure-core/src/cryptography/md5.cpp:73]: I am even more confused now.

The lifecycle of a private member is to init on instance cttor and die on instance dttor. Isn't it?
versus an static global variable which gets init at some point during app start up and gets destroyed at some point on app exit.

So if I do,

main() {

 {
   MyClass instance;
 }
}

The private members from `MyClass` init and die on its scope.

Comment (Jinming-Hu) [reply sdk/core/azure-core/src/cryptography/md5.cpp:73]: Oh, okay, my mistake. I thought it was `static`.

Comment (Jinming-Hu) [reply sdk/core/azure-core/src/cryptography/md5.cpp:73]: I'll mark this conversation as resolved.

Comment (Jinming-Hu) [inline sdk/core/azure-core/src/cryptography/md5.cpp:73]: I've got another question now. Are we going to create a new `AlgorithmProviderInstance` instance for each MD5 calculation?

File: sdk/core/azure-core/CMakeLists.txt (hunk 1 above)
Comment (vhvb1989) [reply sdk/core/azure-core/CMakeLists.txt:52]: Actually, there is only one change to the header for MD5 class.. other than that, it's the same.

Comment (vhvb1989) [reply sdk/core/azure-core/CMakeLists.txt:52]: So, it might not be that hard to let just one header.

File: sdk/core/azure-core/src/cryptography/md5.cpp (hunk 2 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/cryptography/md5.cpp:73]: Yes, there are two size_t fields in it and a handle.  There's not a super difference from using only one.
The benefit, in the other hand, of letting each md5 instance to have it's own is to avoid someone updating the global instance at some point and changing all md5 instances right?  For instance, if someone re-asign it for SHA or something else, all md5 instances would react to it.  It would be a weird edge case, but we can prevent it.

Comment (Jinming-Hu) [reply sdk/core/azure-core/src/cryptography/md5.cpp:73]: This is not good. The old impl is all MD5 instances share the same algorithm handle, because in the API reference it says

```
the BCryptOpenAlgorithmProvider function is a relatively time intensive function. Because of this, we recommend that you cache any algorithm provider handles that you will use more than once, rather than opening and closing the algorithm providers over and over.
```

[link](https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptopenalgorithmprovider)

Comment (Jinming-Hu) [reply sdk/core/azure-core/src/cryptography/md5.cpp:73]: > For instance, if someone re-asign it for SHA or something else, all md5 instances would react to it.

Different algorithm will have different static handles, so that they won't interfere with each other.

Comment (vhvb1989) [reply sdk/core/azure-core/src/cryptography/md5.cpp:73]: oh crap. Well, I will update it to use only one provider then. Thank you @Jinming-Hu

File: sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp [hunk 3]
```diff
@@ -135,7 +136,7 @@ namespace Azure { namespace Core { namespace Cryptography {
     ~Md5Hash() override;
 
   private:
-    void* m_md5Context;
+    std::unique_ptr<Hash> m_implementation;
```
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp:139]: It works, but is it a common design pattern?

**Review by Jinming-Hu — APPROVED** [review]

File: sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp (hunk 3 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp:139]: Well, @CaseyCarter  might comment about it.  I don't really know if that's a pattern.

I just wanted to avoid using void* and all the dynamic_casting by introducing a second Abstraction layer for the OS

File: sdk/core/azure-core/src/cryptography/md5.cpp [hunk 4]
```diff
@@ -16,100 +16,71 @@
...
-    };
-
-    struct Md5HashContext
-    {
-      std::string buffer;
-      BCRYPT_HASH_HANDLE hashHandle = nullptr;
-      std::size_t hashLength = 0;
-    };
-  } // namespace _detail
+struct Md5AlgorithmProvider
```
Comment (RickWinter) [inline sdk/core/azure-core/src/cryptography/md5.cpp:23]: This should be a class per our design patterns.

File: sdk/core/azure-core/src/cryptography/md5.cpp [hunk 5]
```diff
@@ -16,100 +16,71 @@
...
+        Handle,
+        BCRYPT_OBJECT_LENGTH,
+        reinterpret_cast<PBYTE>(&objectLength),
+        sizeof(objectLength),
+        &dataLength,
         0);
     if (!BCRYPT_SUCCESS(status))
     {
-      throw std::runtime_error("BCryptCreateHash failed");
+      throw std::runtime_error("BCryptGetProperty failed");
```
Comment (RickWinter) [inline sdk/core/azure-core/src/cryptography/md5.cpp:47]: As a developer/customer what would I do when I see this error?  I suggest adding in the status code or expanding the status code into text to help diagnose the failure.

File: sdk/core/azure-core/src/cryptography/md5.cpp [hunk 6]
```diff
@@ -16,100 +16,71 @@
...
+    status = BCryptGetProperty(
+        Handle,
+        BCRYPT_HASH_LENGTH,
+        reinterpret_cast<PBYTE>(&hashLength),
+        sizeof(hashLength),
+        &dataLength,
+        0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("BCryptGetProperty failed");
```
Comment (RickWinter) [inline sdk/core/azure-core/src/cryptography/md5.cpp:60]: Ditto on the status code for debuggability.

File: sdk/core/azure-core/src/cryptography/md5.cpp [hunk 7]
```diff
@@ -119,53 +90,88 @@ namespace Azure { namespace Core { namespace Cryptography {
...
+        GetMD5AlgorithmProvider().Handle,
+        &m_hashHandle,
+        reinterpret_cast<PUCHAR>(&m_buffer[0]),
+        static_cast<ULONG>(m_buffer.size()),
+        nullptr,
+        0,
+        0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("BCryptCreateHash failed");
```
Comment (RickWinter) [inline sdk/core/azure-core/src/cryptography/md5.cpp:124]: Ditto on the status code

Comment (RickWinter) [reply sdk/core/azure-core/src/cryptography/md5.cpp:124]: Given there are multiple places that do this, Add a ThrowIfFailed(status) method.

File: sdk/core/azure-core/src/cryptography/md5.cpp [hunk 8]
```diff
@@ -119,53 +90,88 @@ namespace Azure { namespace Core { namespace Cryptography {
...
+    }
+  }
+  ~Md5BCrypt() {}
```
Comment (RickWinter) [inline sdk/core/azure-core/src/cryptography/md5.cpp:127]: You need to release the `BCRYPT_HASH_HANDLE` by calling `BCryptDestroyHash`

**Review by RickWinter — CHANGES_REQUESTED** [review]

File: sdk/core/azure-core/src/cryptography/md5.cpp (hunk 4 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/cryptography/md5.cpp:23]: You mean this guideline:
https://azure.github.io/azure-sdk/cpp_implementation.html#cpp-design-naming-classes-no-struct-keyword
???

This class has only public members, it's following the guideline

File: sdk/core/azure-core/test/ut/md5.cpp [hunk 9]
```diff
@@ -98,7 +100,7 @@ TEST(Md5Hash, Basic)
 TEST(Md5Hash, ExpectThrow)
 {
   std::string data = "";
-  const uint8_t* ptr = reinterpret_cast<const uint8_t*>(data.data());
+  const uint8_t* ptr = reinterpret_cast<const uint8_t*>(data.c_str());
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/md5.cpp:103]: I am guessing this is to get a null-terminated value, which is good. But did we need the change for some particular reason?

File: sdk/core/azure-core/src/cryptography/md5.cpp [hunk 10]
```diff
@@ -16,156 +16,195 @@
...
 
 #if defined(AZ_PLATFORM_WINDOWS)
 
-  namespace _detail {
-    struct AlgorithmProviderInstance
+void ThrowIfFails(NTSTATUS status)
+{
+  if (!NT_SUCCESS(status))
+  {
+    throw std::runtime_error("BCryptOpenAlgorithmProvider failed");
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/cryptography/md5.cpp:27]: Append the status value to the message.

File: sdk/core/azure-core/src/cryptography/md5.cpp [hunk 11]
```diff
@@ -16,156 +16,195 @@
...
+public:
+  BCRYPT_ALG_HANDLE Handle;
+  std::size_t ContextSize;
+  std::size_t HashLength;
+
+  Md5AlgorithmProvider()
+  {
+    // open an algorithm handle
+    if (!NT_SUCCESS(
+            m_status = BCryptOpenAlgorithmProvider(&Handle, BCRYPT_MD5_ALGORITHM, nullptr, 0)))
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/cryptography/md5.cpp:44]: Can we use `ThrowIfFails` here and elsewhere?

File: sdk/core/azure-core/src/cryptography/md5.cpp [hunk 12]
```diff
@@ -16,156 +16,195 @@
...
-    MD5_Update(md5Context, data, length);
+    MD5_Update(m_context.get(), data, length);
   }
 
-  std::vector<uint8_t> Md5Hash::OnFinal(const uint8_t* data, std::size_t length)
+  std::vector<uint8_t> OnFinal(const uint8_t* data, std::size_t length)
   {
     OnAppend(data, length);
-    MD5_CTX* md5Context = static_cast<MD5_CTX*>(m_md5Context);
     unsigned char hash[MD5_DIGEST_LENGTH];
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/cryptography/md5.cpp:180]: This is only 16 bytes, let's save this buffer as a private field.

File: sdk/core/azure-core/src/cryptography/md5.cpp [hunk 13]
```diff
@@ -16,156 +16,195 @@
...
+    MD5_Init(m_context.get());
+  }
+};
+
+} // namespace
+Azure::Core::Cryptography::Md5Hash::Md5Hash() : m_implementation(std::make_unique<Md5OpenSSL>()) {}
 #endif
+
+namespace Azure { namespace Core { namespace Cryptography {
+  Md5Hash::~Md5Hash() {}
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/cryptography/md5.cpp:198]: What's this supposed to do?

File: sdk/core/azure-core/src/cryptography/md5.cpp [hunk 14]
```diff
@@ -16,156 +16,195 @@
...
 
-  void Md5Hash::OnAppend(const uint8_t* data, std::size_t length)
+} // namespace
+Azure::Core::Cryptography::Md5Hash::Md5Hash() : m_implementation(std::make_unique<Md5BCrypt>()) {}
+
+#elif defined(AZ_PLATFORM_POSIX)
+
+class Md5OpenSSL : public Azure::Core::Cryptography::Hash {
+private:
+  std::unique_ptr<MD5_CTX> m_context;
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/cryptography/md5.cpp:170]: Do we need a dtor to delete this `m_context`, or will the unique ptr help us here and clean up as soon as `Md5OpenSSL` is out of scope?

File: sdk/core/azure-core/test/ut/md5.cpp [hunk 15]
```diff
@@ -118,3 +120,32 @@ TEST(Md5Hash, CtorDtor)
     Md5Hash instance;
   }
 }
+
+TEST(Md5Hash, multiThread)
+{
+  auto hashThreadRoutine = [](int sleepFor) {
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/md5.cpp:126]: Any bug fix in this PR that would be good to call out in the CL?

File: sdk/core/azure-core/test/ut/md5.cpp (hunk 9 above)
Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/md5.cpp:103]: .data returns the ptr to the start without any warranty about all the content is contiguous, while c_str makes sure that the ptr has an address to the all the string content up to the null-ending.

std::string can use non-contiguos memory to hold the string

File: sdk/core/azure-core/src/cryptography/md5.cpp (hunk 12 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/cryptography/md5.cpp:180]: Can be done later if we need for such a thing.  Not relevant for now

File: sdk/core/azure-core/src/cryptography/md5.cpp (hunk 14 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/cryptography/md5.cpp:162]: > will the unique ptr help us here and clean up

Yes

File: sdk/core/azure-core/src/cryptography/md5.cpp (hunk 13 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/cryptography/md5.cpp:190]: Defines the dttor as no-op.
The base class `Hash` defines a virtual dttor `virtual ~Hash() = default;` which obligated derived classes to define how the class is destroyed.
For the top level abstraction, we don't care about how it is destoyed because it only wraps a second layer abstraction which depends on the OS and where the dttor matters.

File: sdk/core/azure-core/test/ut/md5.cpp (hunk 15 above)
Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/md5.cpp:126]: there's no bug fix.
The previous impl was thread safe as well.

File: sdk/core/azure-core/src/cryptography/md5.cpp (hunk 7 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/cryptography/md5.cpp]: something like `ThrowIfFailed` would require to know what message to display next to the `status` code before even knowing if we are throwing or not.

I have updated the code adding the code

File: sdk/core/azure-core/src/cryptography/md5.cpp (hunk 11 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/cryptography/md5.cpp]: Not if we want to display a message next to the code.  Because the message would need to be created even before we throw

File: sdk/core/azure-core/src/cryptography/md5.cpp (hunk 12 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/cryptography/md5.cpp:172]: Do repeated calls to Final fail in the exact same way for both implementations/platforms? We want the behavior to be consistent. If it is consistent (including the error condition), then you are right, we can do it later. That said, one of the goals of the issue was to make sure we have defined behavior for these edge cases.

Comment (ahsonkhan) [reply sdk/core/azure-core/src/cryptography/md5.cpp:172]: Oh nevermind, the base class is guaranteeing that:
https://github.com/Azure/azure-sdk-for-cpp/blob/b947735b2262beeb9b0c731ba6dd224019a02b22/sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp#L91-L95

File: sdk/core/azure-core/src/cryptography/md5.cpp [hunk 16]
```diff
@@ -16,156 +16,187 @@
...
-namespace Azure { namespace Core { namespace Cryptography {
+namespace {
 
 #if defined(AZ_PLATFORM_WINDOWS)
 
-  namespace _detail {
-    struct AlgorithmProviderInstance
+class Md5AlgorithmProvider {
+private:
+  NTSTATUS m_status = 0;
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/cryptography/md5.cpp:25]: Why is the default value 0? Maybe add a comment. Same with `Md5BCrypt` below

File: sdk/core/azure-core/src/cryptography/md5.cpp [hunk 17]
```diff
@@ -16,156 +16,187 @@
...
+    hash.resize(m_hashLength);
+    if (!BCRYPT_SUCCESS(
+            m_status = BCryptFinishHash(
+                m_hashHandle,
+                reinterpret_cast<PUCHAR>(&hash[0]),
+                static_cast<ULONG>(hash.size()),
+                0)))
     {
-      throw std::runtime_error("BCryptFinishHash failed");
+      throw std::runtime_error("BCryptFinishHash failed with code" + m_status);
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/cryptography/md5.cpp:121]: needs some spacing or punctuation so it doesn't read like: `BCryptFinishHash failed with code4`

Same issue everywhere else.

```suggestion
      throw std::runtime_error("BCryptFinishHash failed with code: " + m_status + ".");
```

**Review by ahsonkhan — APPROVED** [review]
Fixup the error messages, otherwise looks good.

**Review by RickWinter — APPROVED** [review]

File: sdk/core/azure-core/test/ut/md5.cpp (hunk 9 above)
Comment (CaseyCarter) [reply sdk/core/azure-core/test/ut/md5.cpp:103]: > std::string can use non-contiguos memory to hold the string

This hasn't been true since C++11. The only difference is that `data.data()` returns a pointer to `char` (since `data` is not `const`) whereas `data.c_str()` returns a pointer to `const char`.

Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/md5.cpp:103]: Thank you @CaseyCarter , so, I am curious.  What would it happen if I keep appending data to one string until all the reserved contiguous memory is full?  On the next append, string would need to move all the data to a new bigger contiguous memory right?  If that happen, would the return value of `data()` and `c_str()` changed to a new address? 
And the previous result of calling `c_str()`, what happen if I tried to access it latter after the string kept growing and at some point moved?  (classic segmentation fault?)

Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/md5.cpp:103]: also @CaseyCarter, I was learning about the concept `trivially copyable`, and I was understood that only contiguous-memory objects are trivially copyable, and string returns false when checking if it is trivially copyable.  Now I am confused, is that only for backward compatibility to before C++11?

Comment (CaseyCarter) [reply sdk/core/azure-core/test/ut/md5.cpp:103]: > Thank you @CaseyCarter , so, I am curious. What would it happen if I keep appending data to one string until all the reserved contiguous memory is full? On the next append, string would need to move all the data to a new bigger contiguous memory right? If that happen, would the return value of `data()` and `c_str()` changed to a new address?

Yes; when the size of a `vector`/`string` needs to exceed the current capacity it allocates a buffer whose capacity is some multiple of the current capacity. (For MSVC, "some multiple" is 3/2, I think other standard libraries use 2.) This reallocation invalidates all pointers/references/iterators previously obtained from that `vector`/`string`, including the results of `data` and `c_str`. 

> And the previous result of calling `c_str()`, what happen if I tried to access it latter after the string kept growing and at some point moved? (classic segmentation fault?)

It's undefined behavior so in theory anything can happen; in practice, use-after-free usually manifests as a segmentation fault or (the worst possible result) just observes garbage values in the memory.

> string returns false when checking if it is trivially copyable.

For a trivially copyable type, it's possible to simply copy the bytes of an object's representation into the bytes of another object of the same type; you're guaranteed that the result is still a valid object and that it has the same value as the source object. The elements of `std::string` are trivially copyable (which generalizes to other specializations of `basic_string` like `std::wstring` and `std::u8string`) but strings themselves can't be trivially copyable because they manage memory. If you copy the bytes of the representation of a `std::string` into another `std::string` what you'll get is a memory leak and a double-free.
